File: vbnet/list_merger/list_merger_2.vb
' This program is called List Merger.
' The purpose of this program is to merge two lists of integers into a single list.
' It demonstrates basic list operations such as adding elements and iterating through lists.
' The program is designed to be verbose with comments to help new programmers understand each step.

Module ListMerger

    Sub Main()
        ' Declare and initialize the first list of integers
        Dim list1 As New List(Of Integer)
        list1.Add(1)
        list1.Add(2)
        list1.Add(3)

        ' Declare and initialize the second list of integers
        Dim list2 As New List(Of Integer)
        list2.Add(4)
        list2.Add(5)
        list2.Add(6)

        ' Declare the merged list of integers
        Dim mergedList As New List(Of Integer)

        ' Add elements from the first list to the merged list
        For Each item In list1
            mergedList.Add(item)
        Next

        ' Add elements from the second list to the merged list
        For Each item In list2
            mergedList.Add(item)
        Next

        ' Print the merged list to the console
        Console.WriteLine("Merged List:")
        For Each item In mergedList
            Console.WriteLine(item)
        Next

        ' Keep the console window open
        Console.ReadLine()
    End Sub

End Module

File: c/list_merger/list_merger_2.c
#include <stdio.h>
#include <stdlib.h>

// This program merges two sorted linked lists into a single sorted linked list.
// It demonstrates the use of dynamic memory allocation, linked list operations,
// and proper memory management practices. The program is designed to help 
// programmers understand how to work with linked lists and dynamic memory in C.

typedef struct Node {
    int data;
    struct Node* next;
} Node;

// Function to create a new node with given data
Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

// Function to merge two sorted linked lists
Node* mergeLists(Node* list1, Node* list2) {
    Node* mergedList = NULL;
    Node** lastPtrRef = &mergedList;

    while (list1 != NULL && list2 != NULL) {
        if (list1->data <= list2->data) {
            *lastPtrRef = list1;
            list1 = list1->next;
        } else {
            *lastPtrRef = list2;
            list2 = list2->next;
        }
        lastPtrRef = &((*lastPtrRef)->next);
    }

    if (list1 != NULL) {
        *lastPtrRef = list1;
    } else {
        *lastPtrRef = list2;
    }

    return mergedList;
}

// Function to print the linked list
void printList(Node* head) {
    Node* current = head;
    while (current != NULL) {
        printf("%d -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");
}

// Function to free the linked list
void freeList(Node* head) {
    Node* current = head;
    Node* nextNode;
    while (current != NULL) {
        nextNode = current->next;
        free(current);
        current = nextNode;
    }
}

int main() {
    // Creating first sorted linked list: 1 -> 3 -> 5
    Node* list1 = createNode(1);
    list1->next = createNode(3);
    list1->next->next = createNode(5);

    // Creating second sorted linked list: 2 -> 4 -> 6
    Node* list2 = createNode(2);
    list2->next = createNode(4);
    list2->next->next = createNode(6);

    // Merging the two sorted linked lists
    Node* mergedList = mergeLists(list1, list2);

    // Printing the merged linked list
    printList(mergedList);

    // Freeing the merged linked list
    freeList(mergedList);

    return 0;
}

